inference.py

```diff
@@ -1,61 +1,107 @@
     task_ids = list(problems.keys())
+    
+    return prompts, task_ids
 
+def run_generation(out_file, config):
+    sampling_params = config["sampling"]
+    llm_params = config["llm_params"]
+    prompts, task_ids = get_prompts()
+    
     llm = LLM(**llm_params)
     t0 = time()
     outputs = llm.generate(prompts, SamplingParams(**sampling_params))
-    print(f"generation time: {time() - t0:.3f}")
+    elapsed_time = time() - t0
+    print(f"generation time: {elapsed_time:.3f}")
 
     samples = []
     for tid, output in zip(task_ids, outputs):
         for out in output.outputs:
             samples.append(dict(task_id=tid, completion=out.text))
     write_jsonl(out_file, samples)
+    
+    return elapsed_time
 
+def create_experiment_dir(base_path="/raid/shared/llm-inference-scaling/experiments"):
+    base_path = Path(base_path)
+    experiment_name = '-'.join(coolname.generate())
+    experiment_path = base_path / experiment_name
+    experiment_path.mkdir(exist_ok=True, parents=True)
+    
+    return experiment_name, experiment_path
 
-def run_experiment(sampling_params, llm_params, evaluate=False):
-    environ["CUDA_VISIBLE_DEVICES"] = "3"  # todo do this differently
-    environ["TOKENIZERS_PARALLELISM"] = "true"
-
-    out_path = Path("/raid/shared/llm-inference-scaling/outputs")
-    out_path.mkdir(exist_ok=True, parents=True)
-
-    name = f"{uuid.uuid4()}.jsonl"  # choose out file name randomly
-    this_params = dict(sampling_params=sampling_params, llm_params=llm_params)
-    experiments_file = out_path / "_experiments.json"
+def store_metadata(name, config, base_path="/raid/shared/llm-inference-scaling/experiments"):
+    experiments_file = Path(base_path) / "experiments.json"
     if experiments_file.exists():
         with open(experiments_file, "r") as f:
             experiments = json.load(f)
     else:
         experiments = {}
-    if this_params in experiments.values():
-        return
-    experiments[name] = this_params
-    out_file = out_path / name
-    run_generation(out_file, sampling_params, llm_params)
+    experiments[name] = config
+    
+    return experiments, experiments_file
+    
+def experiment_setup(config):
+    # Choose GPU
+    environ["CUDA_VISIBLE_DEVICES"] = "5"  # todo do this differently
+    environ["TOKENIZERS_PARALLELISM"] = "true"
+    experiment_name, out_path = create_experiment_dir()
+    experiments, experiments_file = store_metadata(experiment_name, config)
+    return experiments, experiments_file, experiment_name, out_path
+    
+def store_results(experiments, experiments_file):
     with open(experiments_file, "w") as f:
-        json.dump(experiments, f, indent=4)  # write only when completed
-        # todo write time taken
-    if evaluate:
-        evaluation.evaluate_functional_correctness(
+        json.dump(experiments, f, indent=4)
+
+def evaluate(out_file):
+    evaluation.evaluate_functional_correctness(
             str(out_file), k=[1, 4, 16, 64, 256]
```

Comment:
What is k?
From looking at it and not knowing much, it might be something that should depend on the config.

Reply:
It's (probably) the points for which humaneval evaluation actually calculates the pass@k and then extrapolates from those. So it should be from 1 to n from config with reasonable steps. Something like powers of 2 as long as they are smaller than n as well as n.